libs/components/config/src/lib/params.spec.ts

```diff
@@ -319,6 +393,6 @@ describe('SkyAppRuntimeConfigParams', () => {
 
     expect(
       params.getLinkUrl('https://mysite.com?foobar=Robert+Hernandez')
-    ).toEqual('https://mysite.com?foobar=Robert+Hernandez&a1=%20');
+    ).toEqual('https://mysite.com?foobar=Robert%2BHernandez&a1=%20');
```

Comment:
This should remain a `+` character; we have a consumer that returns encoded URLs from their backend (replacing spaces with "+") and we were double-encoding the params for them, breaking their service. Here's the original PR: https://github.com/blackbaud/skyux/pull/1462/files

Reply:
We need to clarify the functional requirements of the function.  Currently the behavior of getLinkUrl is inconsistent - it will sometimes encode parms and it will sometimes _not_ encode them.

1) You are correct that we are now treating params provided on the inbound url as something we need to encode, in the same way we are encoding parameters we add.
2) Switching + to %2B isn't double encoding in of itself, but I agree if the caller assumes we aren't doing any encoding this will cause double encoding.
3) Would you expect the encoding behavior to be _different_ depending on how the parameters are ingested?

For example, what should the following do:
`fit('should validate query string encoding', () => {
    const params = new SkyAppRuntimeConfigParams('?a1=str+1', allowed);

    expect(
      params.getLinkUrl('https://mysite.com?a2=str+2', { a3: 'str+3' })
    ).toEqual('____');
  });`

4) SkyHref will immediately set the response to href of the native element.  We need to encode it, or we will generate malformed urls
5) Currently encoding is handled by UrlEncoder defined in params.ts which wraps encodeURIComponent.  Do you want to follow [angular's practice](https://github.com/angular/angular/blob/cb31dbc75ca4141d61cec3ba6e60505198208a0a/packages/common/http/src/params.ts#L96-L111) where we encode using encodeURIComponent and then put back + and any other subset of characters?

Reply:
Params that already exist in the URL should not be encoded, so I would expect this test to pass as-is: https://github.com/blackbaud/skyux/pull/1462/files#diff-07e3b4a4c4dadbafa1bd357b6708e7c9636d97376d5e54d4d90188f545541a26R313

Params that "we own" (added from `SkyuxConfigParams` or from the `queryParams` input) should be encoded, so I would expect your test to output: 
```
https://mysite.com?a2=str+2&a3=str%2B3&a1=str+1
```

Granted, this is not great behavior (ideally all params would be encoded correctly) but our consumers have made some strange decisions and we don't want to break them in a minor version. If you'd like, you could put a `// TODO: encode all params in a future major version` comment above the encoding logic so we remember to come back and consider a breaking change in a major version.

Reply:
I refactored the logic to walk a fine line.  We will respect the current url parameters if they do not require modification, deferring to the caller that they are encoded as expected.  If we need to modify those parameters because they should either be removed or have their values overridden, we will fold the existing parameters into the unified set and encode them properly.